src/java/org/apache/cassandra/db/tries/DeletionAwareTrie.java

```diff
@@ -0,0 +1,419 @@
+        boolean enterDeletionsBranch();
+
+        /// Called for every deletion marker found in the deletion branch.
+        void deletionMarker(D marker);
+
+        /// Called when the deletion branch is exited.
+        void exitDeletionsBranch();
+    }
+
+    @Override
+    default String dump(Function<T, String> contentToString)
+    {
+        return dump(contentToString, Object::toString);
+    }
+
+    default String dump(Function<T, String> contentToString,
+                        Function<D, String> rangeToString)
+    {
+        return process(Direction.FORWARD, new TrieDumper.DeletionAware<>(contentToString, rangeToString));
+    }
+
+    /// Process the trie using the given [DeletionAwareWalker].
+    default <R> R process(Direction direction, DeletionAwareWalker<? super T, ? super D, R> walker)
+    {
+        return cursor(direction).process(walker);
+    }
+
+    /// Returns a view of the live content in this trie as a regular [Trie].
+    default Trie<T> contentOnlyTrie()
+    {
+        return this::cursor;
+    }
+
+    /// Returns a view of all deletion ranges in this trie as a single [RangeTrie].
+    default RangeTrie<D> deletionOnlyTrie()
+    {
+        // Note: We must walk the main trie to find deletion branch roots. This can be inefficient.
+        return dir -> new DeletionAwareCursor.DeletionsTrieCursor<>(cursor(dir));
+    }
+
+    /// Returns a view of the combination of the live data and deletions in this trie as a regular [Trie], using
+    /// the provided mapping function to covert values to a common type.
+    default <Z> Trie<Z> mergedTrie(BiFunction<T, D, Z> resolver)
+    {
+        return dir -> new DeletionAwareCursor.LiveAndDeletionsMergeCursor<>(resolver, cursor(dir));
+    }
+
+    static <T, D extends RangeState<D>>
+    DeletionAwareTrie<T, D> empty(ByteComparable.Version byteComparableVersion)
+    {
+        return direction -> new DeletionAwareCursor.Empty<>(direction, byteComparableVersion);
+    }
+
+    @Override
+    default DeletionAwareTrie<T, D> prefixedBy(ByteComparable prefix)
+    {
+        return dir -> new PrefixedCursor.DeletionAware<>(prefix, cursor(dir));
+    }
+
+    /// @inheritDoc
+    ///
+    /// Note: if the cursor is positioned below a deletion branch root, the tail will not include any information about
+    /// that deletion branch, even if it applies to the current position.
+    @Override
+    default DeletionAwareTrie<T, D> tailTrie(ByteComparable prefix)
+    {
+        DeletionAwareCursor<T, D> c = cursor(Direction.FORWARD);
+        if (c.descendAlong(prefix.asComparableBytes(c.byteComparableVersion())))
+            return c::tailCursor;
+        else
+            return null;
+    }
+
+    /// Returns an entry set containing all tail tree constructed at the points that contain content of
+    /// the given type.
+    default Iterable<Map.Entry<ByteComparable, DeletionAwareTrie<T, D>>> tailTries(Direction direction, Class<? extends T> clazz)
```

Comment:
Seems unused.

Reply:
This will likely be needed, but also uses the implementation that memtable range queries are built on.

Added a test, which uncovered some problems.